### PR TITLE
IGNITE-17258 Implement ReplicaListener

### PR DESCRIPTION
**Pull request by vldpyatkov**

### CODE REVIEW

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 1]
```diff
@@ -177,9 +188,37 @@ public void onReceived(NetworkMessage message, NetworkAddress senderAddr, @Nulla
...
+                    CompletableFuture<Object> result = replica.processRequest(request);
+
+                    result.handle((res, ex) -> {
+                        if (ex != null) {
+                            clusterNetSvc.messagingService().respond(
+                                    senderAddr,
+                                    REPLICA_MESSAGES_FACTORY.replicaResponse().result(res).build(),
+                                    correlationId);
+                        } else {
+                            var traceId = UUID.randomUUID();
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:200]: What do you mean? ex is null here.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 2]
```diff
@@ -177,9 +188,37 @@ public void onReceived(NetworkMessage message, NetworkAddress senderAddr, @Nulla
                         //TODO:IGNITE-17255 Send an exceptional response to the client side when the replica is absent.
                     }
 
-                    ReplicaResponse resp = replica.processRequest(request);
-
-                    clusterNetSvc.messagingService().respond(senderAddr, resp, correlationId);
+                    CompletableFuture<Object> result = replica.processRequest(request);
+
+                    result.handle((res, ex) -> {
+                        if (ex != null) {
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:194]: It's probably if (ex == null) {

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/DualRowReplicaRequest.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Dual row replica request.
+ */
+public interface DualRowReplicaRequest extends ReplicaRequest {
+    @Marshallable
+    BinaryRow binaryRow();
+
+    @Marshallable
+    BinaryRow oldBinaryRow();
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/DualRowReplicaRequest.java:33]: Not sure whether old is proper naming here. It's either first and second rows or SwapRowReplicaRequest with expected and new rows.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/action/RequestType.java [hunk 4]
```diff
@@ -15,32 +15,56 @@
...
+    /** RW get operation. */
+    RW_GET,
+
+    /** RW get all operation. */
+    RW_GET_ALL,
+
+    /** RW delete operation. */
+    RW_DELETE,
+
+    /** RW delete all operation. */
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/action/RequestType.java:33]: Let's be consistent with javadocs and remove them, cause instance names are themselves pretty clear.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 5]
```diff
@@ -0,0 +1,789 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/** Partition replication listener. */
+public class PartitionReplicaListener implements ReplicaListener {
+    /** Logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(PartitionReplicaListener.class);
+
+    /** Primary key id. */
+    public static final UUID INDEX_PK_ID = new UUID(0L, 2L);
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:66]: Let's mark all tmp index related stuff(fields and methods) with todo with https://issues.apache.org/jira/browse/IGNITE-17479

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 6]
```diff
@@ -0,0 +1,789 @@
...
+    @Override
+    public CompletableFuture<Object> invoke(ReplicaRequest request) {
+        if (request instanceof ReadWriteSingleRowReplicaRequest) {
+            return processSingleEntryAction((ReadWriteSingleRowReplicaRequest) request);
+        } else if (request instanceof ReadWriteMultiRowReplicaRequest) {
+            return processMultiEntryAction((ReadWriteMultiRowReplicaRequest) request);
+        } else if (request instanceof ReadWriteDualRowReplicaRequest) {
+            return processTwoEntriesAction((ReadWriteDualRowReplicaRequest) request);
+        } else {
+            return null;
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:118]: Seems that throwing UnsupportedReplicaRequestException fits better here.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 7]
```diff
@@ -0,0 +1,789 @@
...
+    }
+
+    /**
+     * Find out a row id by an index.
+     *
+     * @param indexId Index id.
+     * @param key     Key to find.
+     * @return Value or {@code null} if the key does not determine a value.
+     */
+    private RowId valueByUniqueIndex(@NotNull UUID indexId, BinaryRow key) {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:139]: Method name is confusing, it's either valueByKey or rowIdByKey.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 8]
```diff
@@ -0,0 +1,789 @@
...
+        if (INDEX_PK_ID.equals(indexId)) {
+            return primaryIndex.get(key.keySlice());
+        }
+
+        if (INDEX_SCAN_ID.equals(indexId)) {
+            RowId[] rowIdHolder = new RowId[1];
+
+            mvDataStorage.forEach((rowId, binaryRow) -> {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:147]: Am I right that it's a tmp solution that uses storage as index to retrieve rowId with the help of full scan?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 9]
```diff
@@ -0,0 +1,789 @@
...
+    /**
+     * Find out a row id by an index.
+     *
+     * @param indexId Index id.
+     * @param key     Key to find.
+     * @return Value or {@code null} if the key does not determine a value.
+     */
+    private RowId valueByUniqueIndex(@NotNull UUID indexId, BinaryRow key) {
+        if (INDEX_PK_ID.equals(indexId)) {
+            return primaryIndex.get(key.keySlice());
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:141]: Let's add a todo for index-ready logic.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 10]
```diff
@@ -0,0 +1,789 @@
...
+            case RW_GET_AND_DELETE: {
+                CompletableFuture<RowId> lockFut = takeLocksForDelete(searchKey, indexId, txId);
+
+                return lockFut.thenCompose(lockedRowId -> {
+                    BinaryRow result = lockedRowId != null ? mvDataStorage.read(lockedRowId, txId) : null;
+
+                    CompletableFuture raftFut = lockedRowId != null ? raftClient.run(new DeleteCommand(searchKey, txId)) :
+                            CompletableFuture.completedFuture(null);
+
+                    // TODO: add exception handling, including analyzing whether raftFut result.
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:390]: Please add raft-exception handling, including analyzing whether raftFut result contains exception.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 2 above)
Comment (vldpyatkov) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: Sure, fixed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/DualRowReplicaRequest.java (hunk 3 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/DualRowReplicaRequest.java:33]: Renamed the class name.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/action/RequestType.java (hunk 4 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/action/RequestType.java:33]: Staied all javadoc empty.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 6 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:118]: Done.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 7 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:139]: Renamed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 8 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:151]: It is a solution for scanning a partition, when we do not have index.
I do not know is the scan index will be required in the future, or we will always have the primary one.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 9 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:145]: Added todo.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java]: Added todo.

**Pull request closed without merging**